modules/stochastic_tools/src/surrogates/LibtorchSimpleNNTrainer.C

```diff
@@ -0,0 +1,199 @@
+  params.addParam<std::string>(
+      "filename", "net.pt", "Filename used to output the neural net parameters.");
+  params.addParam<bool>("read_from_file",
+                        false,
+                        "Switch to allow reading old trained neural nets for further training.");
+  params.addParam<Real>("learning_rate", 0.001, "Learning rate (relaxation).");
+  params.addRangeCheckedParam<unsigned int>(
+      "print_epoch_loss",
+      0,
+      "0<=print_epoch_loss",
+      "Epoch training loss printing. 0 - no printing, 1 - every epoch, 10 - every 10th epoch.");
+  params.addParam<unsigned int>(
+      "seed", 11, "Random number generator seed for stochastic optimizers.");
+
+  return params;
+}
+
+LibtorchSimpleNNTrainer::LibtorchSimpleNNTrainer(const InputParameters & parameters)
+  : SurrogateTrainer(parameters),
+    _sampler_row(getSamplerData()),
+    _response(getTrainingData<Real>(getParam<ReporterName>("response"))),
+    _num_batches(getParam<unsigned int>("num_batches")),
+    _num_epocs(getParam<unsigned int>("num_epochs")),
+    _rel_loss_tol(getParam<Real>("rel_loss_tol")),
+    _num_neurons_per_layer(declareModelData<std::vector<unsigned int>>(
+        "num_neurons_per_layer", getParam<std::vector<unsigned int>>("num_neurons_per_layer"))),
+    _num_hidden_layers(
+        declareModelData<unsigned int>("num_hidden_layers", _num_neurons_per_layer.size())),
+    _filename(getParam<std::string>("filename")),
+    _read_from_file(getParam<bool>("read_from_file")),
+    _learning_rate(getParam<Real>("learning_rate")),
+    _print_epoch_loss(getParam<unsigned int>("print_epoch_loss"))
+#ifdef LIBTORCH_ENABLED
+    ,
+    _nn(declareModelData<std::shared_ptr<StochasticTools::LibtorchSimpleNeuralNet>>("nn"))
+#endif
+{
+  // We check if MOOSE is compiled with torch, if not this throws an error
+  StochasticToolsApp::requiresTorch(*this);
+
+#ifdef LIBTORCH_ENABLED
+  // Fixing the RNG seed to make sure every experiment is the same.
+  // Otherwise sampling / stochastic gradient descent would be different.
+  torch::manual_seed(getParam<unsigned int>("seed"));
+#endif
+}
+
+void
+LibtorchSimpleNNTrainer::preTrain()
+{
+  // Resize to number of sample points
+  _flattened_data.resize(_sampler.getNumberOfLocalRows() * _sampler.getNumberOfCols());
+  _flattened_response.resize(_sampler.getNumberOfLocalRows());
+}
+
+void
+LibtorchSimpleNNTrainer::train()
+{
+  unsigned int num_parameters = _sampler.getNumberOfCols();
+
+  for (unsigned int i = 0; i < num_parameters; ++i)
+    _flattened_data[_local_row * num_parameters + i] = _sampler_row[i];
+
+  _flattened_response[_local_row] = _response;
+}
+
+void
+LibtorchSimpleNNTrainer::postTrain()
+{
+  _communicator.allgather(_flattened_data);
+  _communicator.allgather(_flattened_response);
+
+#ifdef LIBTORCH_ENABLED
+
+  // Then, we create and load our Tensors
+  unsigned int n_rows = _flattened_response.size();
+  unsigned int n_cols = _sampler.getNumberOfCols();
```

Comment:
It would be easy to read codes if we rename `n_cols` to `n_features`, and `n_rows` to `n_samples`

I could misunderstand these

Reply:
Sounds good, I'll ad this.

Reply:
I'll do num_inputs and num_samples.